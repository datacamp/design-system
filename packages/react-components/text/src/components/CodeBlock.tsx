import '../sideEffects/injectMonoFontFace';

import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import {
  computeDataAttributes,
  ssrSafeFirstChildSelector,
} from '@datacamp/waffles-utils';
import { css } from '@emotion/core';
import React from 'react';

import codeStyle from '../codeStyle';

export interface CodeBlockProps {
  children: string;
  className?: string;
  dataAttributes?: { [key: string]: string };
}

const preStyle = css({
  backgroundColor: tokens.color.opaque.primaryLightest.value.rgb,
  borderRadius: tokens.radii.small.value,
  boxShadow: tokens.shadow.border.value,
  margin: 0,
  marginTop: tokens.size.space[12].value,
  overflow: 'scroll',
  paddingBottom: tokens.size.space[8].value,
  paddingLeft: tokens.size.space[12].value,
  paddingRight: tokens.size.space[12].value,
  paddingTop: tokens.size.space[8].value,
  [ssrSafeFirstChildSelector]: {
    marginTop: 0,
  },
});

const CodeBlock: React.FC<CodeBlockProps> = ({
  children,
  className,
  dataAttributes,
}) => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <pre className={className} css={preStyle} {...parsedDataAttributes}>
      <code css={codeStyle}>{children}</code>
    </pre>
  );
};

export default CodeBlock;
