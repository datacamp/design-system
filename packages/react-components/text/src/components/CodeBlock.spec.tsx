import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import React from 'react';

import CodeBlock from './CodeBlock';

describe('<CodeBlock />', () => {
  it('renders the text in a <pre> and a <code> element', async () => {
    const exampleText = 'this is some code';
    const { getByText } = await axeRender(<CodeBlock>{exampleText}</CodeBlock>);
    const codeElement = getByText(exampleText) as HTMLElement;
    expect(codeElement.tagName).toEqual('CODE');
    expect((codeElement.parentElement as HTMLElement).tagName).toEqual('PRE');
  });

  it('sets the className on the pre element', async () => {
    const className = 'test-class';
    const { container } = await axeRender(
      <CodeBlock className={className}>example code</CodeBlock>,
    );

    expect(container.firstChild).toHaveClass(className);
  });

  it('sets the dataAttributes on the pre element', async () => {
    const data = { test: 'example' };
    const { container } = await axeRender(
      <CodeBlock dataAttributes={data}>example code</CodeBlock>,
    );

    expect(container.firstChild).toHaveAttribute('data-test', 'example');
  });

  describe('snapshots', () => {
    it('renders', async () => {
      const { container } = await axeRender(
        <CodeBlock>example code</CodeBlock>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
