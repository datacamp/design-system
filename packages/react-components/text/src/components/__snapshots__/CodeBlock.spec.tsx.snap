// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CodeBlock /> snapshots renders 1`] = `
.emotion-0 {
  background-color: rgb(239, 235, 228);
  border: 0;
  border-radius: 4px;
  line-height: 1.5;
  margin: 0;
  overflow: scroll;
  padding-bottom: 8px;
  padding-left: 12px;
  padding-right: 12px;
  padding-top: 8px;
}

*:not(style)~.emotion-0 {
  margin-top: 12px;
}

.emotion-1 {
  border: 0;
  border-radius: 4px;
  color: rgb(5, 25, 45);
  font-family: monospace;
  font-family: JetBrainsMonoNL;
  font-size: 14px;
  font-variant-ligatures: none;
  letter-spacing: 0;
  line-height: 1.5;
  margin: 0;
  padding: 0;
}

<pre
  class="emotion-0"
  tabindex="0"
>
  <code
    class="emotion-1"
  >
    example code
  </code>
</pre>
`;
